### DISCUSSION TITLE
Setting a BC constant value as a function of time

### DISCUSSION THREAD

**Discussion by Khaksarkn** [idaholab/moose#26904] [Q&A General] [answered]
Hi,
I have written a coupledNeumannBC code, part of which goes as follows:

`params.addParam<Real>("temperatureE", 273.0, "environment temperature");`
`_temperatureE(getParam<Real>("temperatureE")),`

`return _test[_i][_qp] * _lambdaE * (_coupledTemperature_val[_qp] - _temperatureE);`

<img width="269" alt="Screenshot 2024-02-27 at 11 31 41" src="https://github.com/idaholab/moose/assets/77813275/fec669c4-9ede-460c-9589-9825b1129e18">

_lambdaE is a constant, 
_temperatureE is also a constant (environment temperature)

Now, I want _temperatureE to be a function of time (Something like a multi-line from data I have). I know how to do it with FunctionDirichletBC or similar. But I am not sure how to do it with this particular code. 
Is there a built-in BC of MOOSE that has a similar function? or a type of function, which I can feed into the _temperatureE from the input file?

This is the input file part I use with this BC:
```
[topT]
    type = TKEnvTempCoupledNeumannBC
    variable = temperature
    lambdaE = 10
    temperatureE = 268.15
    temperature = temperature
    boundary = top
  []
```
I would appreciate your help.

**Comment by GiudGiud** [comment]
Hello

You ll want to make `_temperatureE` a Function.
```
params.addParam<FunctionName>("temperatureE", 273.0, "environment temperature");
```

 in the header
```
  const Function & _temperatureE;
```

in the source
```
_temperatureE(getFunction("temperatureE"))
```
and when using it:
`_temperatureE.value(_t, _current_qp[_qp]);`

**Reply by Khaksarkn** [reply]
Thank you very much, 
I changed it but it gives me a member access error:
```
error: member access into incomplete type 'const Function'
    return _test[_i][_qp] * _lambdaE * (_coupledTemperature_val[_qp] - _temperatureE.value(_t, _current_qp[_qp]));
                                                                                    ^
/Users/kiankhaksar/projects/moose/framework/build/header_symlinks/FunctionInterface.h:20:7: note: forward declaration of 'Function'
class Function;
      ^
1 error generated.




This is how I modified the code:

class TKFunctionEnvTempCoupledNeumannBC : public IntegratedBC
{
    public:
        TKFunctionEnvTempCoupledNeumannBC(const InputParameters & parameters);

        static InputParameters validParams();

    protected:
        virtual Real computeQpResidual() override;

    private:
        Real _lambdaE;

        const Function & _temperatureE;

        const VariableValue & _coupledTemperature_val;
};

And the source:

InputParameters
TKFunctionEnvTempCoupledNeumannBC::validParams()
{
    InputParameters params = IntegratedBC::validParams();

    params.addParam<Real>("lambdaE", 1.0, "value multiplied");
    params.addParam<FunctionName>("temperatureE", 273.0, "environment temperature");

    params.addRequiredCoupledVar("temperature", "value on the BC");
    return params;
}

TKFunctionEnvTempCoupledNeumannBC::TKFunctionEnvTempCoupledNeumannBC(const InputParameters & parameters)
    : IntegratedBC(parameters),
    // getting the required parameters from the input file
      _lambdaE(getParam<Real>("lambdaE")),        //The multiplier
      _temperatureE(getFunction("temperatureE")),//The environment temperature
      _coupledTemperature_val(coupledValue("temperature")) //Coupled variable
{
}

Real
TKFunctionEnvTempCoupledNeumannBC::computeQpResidual()
{
    return _test[_i][_qp] * _lambdaE * (_coupledTemperature_val[_qp] - _temperatureE.value(_t, _current_qp[_qp]));
}
```

**Reply by GiudGiud** [reply]
at the top of the source file
```
#include "Function.h"
```
and the top of the header
```
class Function;
```

**Reply by Khaksarkn** [reply]
Thank you very much.
I changed it and that error is resolved. I'm just not sure if I'm setting the return argument correctly or the part about <class Function;>. Since it gives me this:

error: use of undeclared identifier '_current_qp'
    return _test[_i][_qp] * _lambdaE * (_coupledTemperature_val[_qp] - _temperatureE.value(_t, _current_qp[_qp]));
                                                                                               ^
1 error generated.

Should I declare the _current_qp in the header?
I'm sorry, I am new to these function source codes. I appreciate the help.



I put the code down here just in case:

#pragma once

#include "IntegratedBC.h"

class Function;

class TKFunctionEnvTempCoupledNeumannBC : public IntegratedBC
{
    public:
        TKFunctionEnvTempCoupledNeumannBC(const InputParameters & parameters);

        static InputParameters validParams();

    protected:
        virtual Real computeQpResidual() override;

    private:
        Real _lambdaE;

        const Function & _temperatureE;

        const VariableValue & _coupledTemperature_val;
};


--------------


#include "TKFunctionEnvTempCoupledNeumannBC.h"
#include "Function.h"

registerMooseObject("PolarBearApp", TKFunctionEnvTempCoupledNeumannBC);

InputParameters
TKFunctionEnvTempCoupledNeumannBC::validParams()
{
    InputParameters params = IntegratedBC::validParams();

    params.addParam<Real>("lambdaE", 1.0, "value multiplied");
    params.addParam<FunctionName>("temperatureE", 273.0, "environment temperature");

    params.addRequiredCoupledVar("temperature", "value on the BC");
    return params;
}

TKFunctionEnvTempCoupledNeumannBC::TKFunctionEnvTempCoupledNeumannBC(const InputParameters & parameters)
    : IntegratedBC(parameters),
    // getting the required parameters from the input file
      _lambdaE(getParam<Real>("lambdaE")),        //The multiplier
      _temperatureE(getFunction("temperatureE")),//The environment temperature
      _coupledTemperature_val(coupledValue("temperature")) //Coupled variable
{
}

Real
TKFunctionEnvTempCoupledNeumannBC::computeQpResidual()
{
    return _test[_i][_qp] * _lambdaE * (_coupledTemperature_val[_qp] - _temperatureE.value(_t, _current_qp[_qp]));
}

**Reply by GiudGiud** [reply]
No the attribute must have a different name for BCs
it's not
```
_current_qp[_qp]))
```
look at other BCs maybe
for nodal BCs it s `*_current_node`

**Reply by Khaksarkn** [reply]
Yes! You're right. It seems to be `_q_point[_qp]`
I found it in `FunctionPenaltyDirichletBC`

`return _test[_i][_qp] * _lambdaE * (_coupledTemperature_val[_qp] - _temperatureE.value(_t, _q_point[_qp]));`

It compiles. Does it seem right?

**Reply by GiudGiud** [reply, marked as answer]
seems good to me

**Reply by Khaksarkn** [reply]
Thank you Guillaume for the help. I appreciate it.
